common, db: add network id check to prevent datadir collisions

kyriediculous opened the pull request:
**What does this pull request do? Explain your changes. (required)**
Prevents further datadir collisions by checking the chainID from the provided remote ethereum node against the chainID in the database when the datadir and node were first initiated. 

This prevents the edge case where a node operator changes network without changing the network name or connecting to a network URL that doesn't fit the name. 

**Specific updates (required)**
- Introduced db.NetworkID
- Introduced db.SelectKV
- Introduced db.UpdateKV
- save & compare chainID when starting the node

**How did you test each of these updates (required)**
unit tests

**Does this pull request close any open issues?**
Fixes #1098 

**Checklist:**
- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass


TODO: still uses `backend.NetworkID` , swap it out for `backend.ChainID`

Commit:
common, db: add network id check to prevent datadir collisions

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -293,6 +293,20 @@ func main() {
 	if *network == "offchain" {
 		glog.Infof("***Livepeer is in off-chain mode***")
+
+		expectedChainID, err := dbh.NetworkID()
```
I think we can reduce the amount of duplication in the off-chain vs. on-chain case by extracting some logic into a helper function. Perhaps something like this could work:

```
func checkOrStoreChainID(dbh *common.DB, chainID *big.Int) error {
    expectedChainID, err := dbh.ChainID()
    if err != nil {
        return err
    }
    if expectedChainID == nil {
        // No chainID stored yet
        // Store the provided chainID and skip the check
        if err := dbh.SetChainID(chainID); err != nil {
            return err
        }

        return nil
    }

    if expectedChainID.Cmp(chainID) != 0 {
        // Return error
    }

    return nil
}
```

Then in the off-chain case we can simply pass `big.NewInt(0)` into `checkChainID()` as the chainID and in the on-chain case we can fetch the remote chainID and pass it into `checkChainID()`.

What do you think?

kyriediculous replied:
Yeah sounds good !

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -293,6 +293,20 @@ func main() {
+			dbh.UpdateKV("network", "0")
+		}
+		if expectedChainID != nil && expectedChainID.Int64() != 0 {
+			glog.Errorf("Livepeer running in offchain mode, expecting chainID of 0 but got %v. Did you change networks withour changing the network name or datadir?", expectedChainID)
```
"withour" -> "without"

Also, I feel like we can shorten this message a bit an omit "Livepeer" - what do you think?

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -319,6 +333,28 @@ func main() {
+		if expectedChainID != nil {
+			if expectedChainID.Cmp(chainID) != 0 {
+				glog.Errorf(
+					"ethereum node provided chain ID %v for network '%v', but we expected %v. Did you change networks without changing the network name or dataDir?",
```
"dataDir" -> "datadir" to match error message defined above.

The ordering of the arguments seems off - I think it should be `chainID`, `*network` and then `expectedChainID`

kyriediculous replied:
Yeah caught my eye as well, but restructuring the way we log the error when using a helper as you described above

yondonfu commented on common/db.go:
```diff
@@ -169,8 +170,18 @@ func InitDB(dbPath string) (*DB, error) {
+
 	// updateKV prepared statement
-	stmt, err = db.Prepare("UPDATE kv SET value=?, updatedAt = datetime() WHERE key=?")
+	// stmt, err = db.Prepare("UPDATE kv SET value=?2, updatedAt = datetime() VALUES(?1, ?2) WHERE key=?1")
```
Can remove comment

kyriediculous replied:
Yeah I was trying to do an UPDATE statement with VALUES keyword but I couldn't get it to work so I went for INSERT OR REPLACE instead.

yondonfu replied:
UPDATE is for updating existing records so INSERT OR REPLACE is actually a better fit here since we don't have a default value for `chainID`.

Bump on removing comment.

yondonfu commented on common/db.go:
```diff
@@ -323,6 +337,45 @@ func (db *DB) LastSeenBlock() (*big.Int, error) {
+	return id, nil
+}
+
+func (db *DB) SelectKV(key string) (string, error) {
```
Hm while I see why exporting these methods could be useful, I feel like this is leaking details of the DB to the caller (i.e. that we're using a KV table) and am wary of exposing it/encouraging storing/fetching KV pairs with the caller passing in string keys every time they need to call these methods - feels a bit fragile. Isolating the use of string keys to query the underlying KV table in this file feels safer IMO.

kyriediculous replied:
So in other words we'd only export usage specific methods such as `ChainID()` & `SetChainID()` ?

yondonfu replied:
Yep

kyriediculous commented:
Fixed the requested changes as well as failing test in 5821b09

Also added a test in `./test-args.sh` that checks for an error thrown on datadir collisions when using a custom datadir with a different network. 

Want me to add a test that checks for datadir collissions by using the same network name but a different ID? This would probably require passing in an -ethUrl flag so we can simulate a rinkeby connection with a mainnet chainID

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -293,7 +293,15 @@ func main() {
+		}
+
 	} else {
+		glog.Infof("***Livepeer is connected to the %v network***", *network)
```
I don't think we need this log statement since we already log something similar earlier in the file.

yondonfu commented on common/db_test.go:
```diff
@@ -19,6 +19,87 @@ import (
+	row = dbraw.QueryRow("SELECT value FROM kv WHERE key = 'chainID'")
+	err = row.Scan(&networkID)
+	assert.Nil(err)
+	assert.Equal(expectedChainID, networkID)
```
Let's try to be consistent about using chainID based naming instead of networkID since they technically are different.

yondonfu commented on common/db_test.go:
```diff
@@ -19,6 +19,87 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestUpdateKV_Network(t *testing.T) {
```
Should the name of this test be `TestUpdateKVStore`?

yondonfu commented on common/db_test.go:
```diff
@@ -19,6 +19,87 @@ import (
+	assert.Equal(expectedChainID, networkID)
+}
+
+func TestSelectKV(t *testing.T) {
```
Should the name of this test be `TestSelectKVStore`?

yondonfu commented on common/db_test.go:
```diff
@@ -19,6 +19,87 @@ import (
+	defer dbh.Close()
+	defer dbraw.Close()
+
+	dbh.updateKVStore("chainID", expectedChainID)
```
Can use `SetChainID()` here?

yondonfu commented on cmd/livepeer/livepeer.go:
```diff
@@ -820,3 +838,25 @@ func defaultAddr(addr, defaultHost, defaultPort string) string {
+	}
+
+	if expectedChainID.Cmp(chainID) != 0 {
+		return fmt.Errorf("expecting chainID of %v, but got %v. Did you change networks without changing network name or datadir?")
```
Missing `expectedChainID` and `chainID` as parameters to the error

yondonfu commented on test_args.sh:
```diff
@@ -67,11 +67,17 @@ run_lp -broadcaster -datadir "$CUSTOM_DATADIR"
 kill $pid
 
+# expect error because re-use network name with different chainID
+run_lp -broadcaster -datadir "$CUSTOM_DATADIR" -network mainnet $ETH_ARGS || res=$?
```
I ran this locally and I think you may need to increase SLEEP to something like .5 or else the DB is not initialized in the previous command before the process is killed

yondonfu commented:
> Want me to add a test that checks for datadir collissions by using the same network name but a different ID?

I would suggest just adding a test for when the node has an off-chain DB and connects to a ETH network since that is a distinct case from when the node has an on-chain DB and connects to a ETH network (which we have a test for).

yondonfu commented on common/db_test.go:
```diff
@@ -19,6 +19,89 @@ import (
+	defer dbh.Close()
+	defer dbraw.Close()
+
+	var networkID string
```
Let's call this chainID

yondonfu commented on test_args.sh:
```diff
@@ -19,7 +19,7 @@ cat > $TMPDIR/key <<ETH_KEY
 
 ETH_ARGS="-ethKeystorePath $TMPDIR/key"
-SLEEP=0.25
+SLEEP=0.5
```
Since we're not adding any additional tests in this file, can we just revert all changes? Are any of these changes necessary?

kyriediculous replied:
Oh sorry I forgot to change `SLEEP` back

We need the re-assignment of `CUSTOM_DATA_DIR` else we'll get an error saying that we're expecting a different chain ID

yondonfu commented:
Once #1125 is merged to master, let's rebase so we can pull in the latest go-ethereum version and use `ChainID()` instead of `NetworkID()`

kyriediculous commented:
rebased !

yondonfu reviewed (approved):
LGTM
